pallets/pool-system/src/tranches.rs

```diff
@@ -728,14 +817,26 @@ where
 		I: IntoIterator<Item = W>,
 	{
 		let mut res = Vec::with_capacity(self.tranches.deref().len());
-		let iter = self.tranches.deref().iter().zip(with.into_iter());
-
-		for (tranche, w) in iter {
-			let r = f(tranche, w)?;
-			res.push(r);
+		let mut tranche_slice = self.residual_top_slice().iter();
+		let mut with_iter = with.into_iter();
+
+		for _ in 0..self.tranches.len() {
+			match (tranche_slice.next(), with_iter.next()) {
+				(Some(tranche), Some(w)) => res.push(f(tranche, w)?),
+				_ => {
+					return Err(DispatchError::Other(
+						"Tranches struct contains more tranches than combining iterable",
+					))
+				}
+			};
 		}
 
-		Ok(res)
+		match (tranche_slice.next(), with_iter.next()) {
+			(None, None) => Ok(res),
+			_ => Err(DispatchError::Other(
+				"Iterable contains more elements than Tranches tranche count",
+			)),
+		}
```

Comment:
Maybe encapsulate this? It is used in more than 3 places.

Reply:
Definitely! I ended up adding two generic builders for all `combine_*` functions in `Tranches` and `EpocExecutionTranches`, see 25e9a8fd1f5ddc332c09f18fecc66a849735f9a6

Reply:
Nice!